src/hooks/useProductFetch.ts

```diff
@@ -0,0 +1,57 @@
+import { useState, useEffect } from 'react';
+import { fetchProducts } from '../api';
+import { ProductType } from '../types';
+import { formattedKey } from './useProducts.util';
+import { AFTER_FETCH_SIZE, FIRST_FETCH_PAGE, FIRST_FETCH_SIZE } from '../constant/products';
+import { useToast } from './useToast';
+import { ERROR } from '../constant/message';
+
+interface Props {
+  selectBarCondition: Record<string, string>;
+}
+
+export function useProductFetch({ selectBarCondition }: Props) {
+  const [products, setProducts] = useState<ProductType[]>([]);
+  const [page, setPage] = useState(FIRST_FETCH_PAGE);
+  const [hasMore, setHasMore] = useState(true);
+  const [isLoading, setIsLoading] = useState(false);
+  const { showToast } = useToast();
+
+  useEffect(() => {
+    setProducts([]);
+    setPage(FIRST_FETCH_PAGE);
+    setHasMore(true);
+  }, [selectBarCondition.category, selectBarCondition.sort]);
+
+  useEffect(() => {
+    const setFetchedProducts = async () => {
+      try {
+        setIsLoading(true);
+
+        const size = page === FIRST_FETCH_PAGE ? FIRST_FETCH_SIZE : AFTER_FETCH_SIZE;
+        const queryPage =
+          page === FIRST_FETCH_PAGE
+            ? FIRST_FETCH_PAGE
+            : FIRST_FETCH_SIZE / AFTER_FETCH_SIZE - 1 + page;
+        const newProducts = await fetchProducts({
+          page: queryPage,
+          size,
+          category: selectBarCondition.category,
+          sort: formattedKey(selectBarCondition.sort),
+        });
+        setProducts((prev) => [...prev, ...newProducts]);
+        setHasMore(newProducts.length > 0);
+      } catch (error) {
```

Comment:
현재 이 로직에서 page가 0이 되면서 한번 더 로직이 실행돼요~ 다음과 같은 케이스를 재현해보세요
1. 전체 - 높은 가격순으로 정렬
2. 스크롤을 내려 이후 페이지 로딩
3. 전체 - 낮은 가격순으로 정렬
4. 처음 4개의 아이템이 이전 옵션에서 남아서 넘어옴

![image](https://github.com/Jaymyong66/react-shopping-products/assets/85233397/8829923d-947f-42b2-a04d-52f6481a5b5c)

fetch되기전에 page가 0이 아닐때의 4개 값이 들어오고, 이후에 20개의 요청을 보내서 4+20개로 덮어씌워 지고 있어요
급하게 해결할 수 있는 방법으로는 아래와 같은 방법이 있겠네요~
```suggestion
  useEffect(() => {
    const setFetchedProducts = async () => {
      try {
        setIsLoading(true);

        const size = page === FIRST_FETCH_PAGE ? FIRST_FETCH_SIZE : AFTER_FETCH_SIZE;
        const queryPage =
          page === FIRST_FETCH_PAGE
            ? FIRST_FETCH_PAGE
            : FIRST_FETCH_SIZE / AFTER_FETCH_SIZE - 1 + page;
        const newProducts = await fetchProducts({
          page: queryPage,
          size,
          category: selectBarCondition.category,
          sort: formattedKey(selectBarCondition.sort),
        });
        setProducts((prev) => {
          return page === 0 ? newProducts : [...prev, ...newProducts];
        });
        setHasMore(newProducts.length > 0);
      } catch (error) {
```

Reply:
와 토다리한테 맨날 QC 받고 싶다..... 감사합니다!!!!!!!!!!